src/devices/mmio.rs

```diff
@@ -0,0 +1,199 @@
+    *v = (*v & !0xffff_ffff) | u64::from(lo)
+}
+
+/// A common interface for Virtio devices that use the MMIO transport, which also provides a
+/// default implementation of read and write operations from/to the device registers and
+/// configuration space.
+// Adding the `M` generic parameter that's also required by `VirtioDevice` for the time being.
+pub trait VirtioMmioDevice<M: GuestAddressSpace>: WithDriverSelect<M> {
+    /// Callback invoked when the driver writes a value to the Queue Notify configuration register.
+    ///
+    /// This is the simplest mechanism the driver can use to notify a virtio MMIO device. The
+    /// meaning of the value is interpreted as specified by the standard. Many VMMs use something
+    /// like the KVM `ioeventfd` mechanism, which eliminates the need to implement this method.
+    // TODO: Move this to the `VirtioDevice` interface if it uniformly applies to all transports.
+    fn queue_notify(&mut self, _val: u32) {
+        // Do nothing by default.
+    }
+
+    /// Handle a driver read operation from the MMIO space of the device.
+    fn read(&mut self, offset: u64, data: &mut [u8]) {
+        match offset {
+            // The standard specifies that accesses to configuration registers are 32-bit wide.
+            0x00..=0xff if data.len() == 4 => {
+                let v = match offset {
+                    0x0 => MMIO_MAGIC_VALUE,
+                    0x04 => MMIO_VERSION,
+                    0x08 => self.device_type(),
+                    0x0c => VENDOR_ID,
+                    0x10 => self.device_features(self.device_features_select()),
+                    0x34 => self
+                        .selected_queue()
+                        .map(Queue::max_size)
+                        .unwrap_or(0)
+                        .into(),
+                    0x44 => self
+                        .selected_queue()
+                        .map(|q| q.ready)
+                        .unwrap_or(false)
+                        .into(),
+                    0x60 => self.interrupt_status().load(Ordering::SeqCst).into(),
+                    0x70 => self.device_status().into(),
+                    0xfc => self.config_generation().into(),
+                    _ => {
+                        warn!("unknown virtio mmio register read: 0x{:x}", offset);
+                        return;
+                    }
+                };
+                // This cannot panic, because we checked that `data.len() == 4`.
+                data.copy_from_slice(v.to_le_bytes().as_ref());
+            }
+
+            // TODO: The standard specifies that configuration space size is device specific, so
+            // we might want to express that via the trait instead of hard coding the current
+            // arbitrary ceiling.
+            // It's ok to use `as` here because `offset` always fits into an `usize` in this case.
+            0x100..=0xfff => self.read_config(offset as usize - 0x100, data),
```

Comment:
According to the spec: `Device-specific configuration space starts at the offset 0x100 and is accessed with byte alignment.`
Admittedly, `byte alignment` is a weird phrase but looking next door, qemu interprets this as `only byte, word or dword accesses are allowed` ([code](https://github.com/qemu/qemu/blob/c8e5c4b246584da36694a3c259a7dbb8a7e7b1f3/hw/virtio/virtio-mmio.c#L113)). I think this makes sense and we should do the same check, WDYT?

Reply:
The byte alignment here means you're not restricted to only accessing it at a word,dword or qword granularity. i.e. you can access it however you like so there is no check you can make.